File: android/sdk/src/main/kotlin/com/clevertap/android/nativedisplay/renderer/BackgroundRenderer.kt
package com.clevertap.android.nativedisplay.renderer

import androidx.compose.animation.core.*
import androidx.compose.foundation.background
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.*
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import coil.compose.rememberAsyncImagePainter
import coil.request.ImageRequest
import com.clevertap.android.nativedisplay.internal.ImageLoaderProvider
import com.clevertap.android.nativedisplay.models.*
import kotlin.math.cos
import kotlin.math.sin
import kotlin.random.Random

/**
 * Apply background to a modifier based on the Background configuration.
 * 
 * This is @Composable because animated backgrounds (Shimmer, AnimatedGradient, Pulse, 
 * Particles) require Compose animation APIs. Static backgrounds (Solid, LinearGradient, etc.) 
 * are delegated to non-composable implementations for optimal performance.
 * 
 * @param background The background configuration to apply
 * @return Modified Modifier with background applied
 */
@Composable
internal fun Modifier.applyBackground(background: Background?): Modifier {
    if (background == null) return this
    
    return when (background) {
        // Static backgrounds - use non-composable implementations
        is Background.Solid -> this.applyStaticSolid(background)
        is Background.LinearGradient -> this.applyStaticLinearGradient(background)
        is Background.RadialGradient -> this.applyStaticRadialGradient(background)
        is Background.SweepGradient -> this.applyStaticSweepGradient(background)
        is Background.Pattern -> this.applyStaticPattern(background)
        is Background.Image -> this.applyStaticImage(background)
        
        // Animated backgrounds - require @Composable
        is Background.Shimmer -> this.applyAnimatedShimmer(background)
        is Background.AnimatedGradient -> this.applyAnimatedGradient(background)
        is Background.Pulse -> this.applyAnimatedPulse(background)
        is Background.Particles -> this.applyAnimatedParticles(background)
        is Background.Layered -> this.applyLayeredBackground(background)
    }
}

// ============================================================================
// STATIC BACKGROUNDS (Non-Composable)
// ============================================================================

/**
 * Apply solid color background (non-composable).
 */
private fun Modifier.applyStaticSolid(bg: Background.Solid): Modifier {
    val color = parseColor(bg.color) ?: Color.Transparent
    return this.background(color)
}

/**
 * Apply linear gradient background (non-composable).
 * Uses drawWithContent so gradient start/end are computed from the actual component
 * size at draw time — fixes solid-color appearance on small elements like buttons.
 */
private fun Modifier.applyStaticLinearGradient(bg: Background.LinearGradient): Modifier {
    val colors = bg.colors.mapNotNull { parseColor(it) }
    if (colors.isEmpty()) return this

    return this.drawWithContent {
        val angleRadians = Math.toRadians((bg.angle - 90.0))
        val start = Offset(
            x = (0.5f - cos(angleRadians).toFloat() * 0.5f) * size.width,
            y = (0.5f - sin(angleRadians).toFloat() * 0.5f) * size.height
        )
        val end = Offset(
            x = (0.5f + cos(angleRadians).toFloat() * 0.5f) * size.width,
            y = (0.5f + sin(angleRadians).toFloat() * 0.5f) * size.height
        )
        val brush = if (bg.stops != null && bg.stops.isNotEmpty()) {
            val colorStops = colors.mapIndexed { index, color ->
                val stop = bg.stops.getOrNull(index) ?: (index.toFloat() / (colors.size - 1))
                stop to color
            }
            Brush.linearGradient(
                colorStops = colorStops.toTypedArray(),
                start = start,
                end = end
            )
        } else {
            Brush.linearGradient(
                colors = colors,
                start = start,
                end = end
            )
        }
        drawRect(brush)
        drawContent()
    }
}

/**
 * Apply radial gradient background (non-composable).
 */
private fun Modifier.applyStaticRadialGradient(bg: Background.RadialGradient): Modifier {
    val colors = bg.colors.mapNotNull { parseColor(it) }
    if (colors.isEmpty()) return this
    
    return this.drawWithContent {
        val brush = if (bg.stops != null && bg.stops.isNotEmpty()) {
            val colorStops = colors.mapIndexed { index, color ->
                val stop = bg.stops.getOrNull(index) ?: (index.toFloat() / (colors.size - 1))
                stop to color
            }
            Brush.radialGradient(
                colorStops = colorStops.toTypedArray(),
                center = Offset(size.width * bg.centerX, size.height * bg.centerY),
                radius = size.minDimension * bg.radius
            )
        } else {
            Brush.radialGradient(
                colors = colors,
                center = Offset(size.width * bg.centerX, size.height * bg.centerY),
                radius = size.minDimension * bg.radius
            )
        }
        drawRect(brush)
        drawContent()
    }
}

/**
 * Apply sweep/conic gradient background (non-composable).
 */
private fun Modifier.applyStaticSweepGradient(bg: Background.SweepGradient): Modifier {
    val colors = bg.colors.mapNotNull { parseColor(it) }
    if (colors.isEmpty()) return this
    
    return this.drawWithContent {
        val brush = if (bg.stops != null && bg.stops.isNotEmpty()) {
            val colorStops = colors.mapIndexed { index, color ->
                val stop = bg.stops.getOrNull(index) ?: (index.toFloat() / (colors.size - 1))
                stop to color
            }
            Brush.sweepGradient(
                colorStops = colorStops.toTypedArray(),
                center = Offset(size.width * bg.centerX, size.height * bg.centerY)
            )
        } else {
            Brush.sweepGradient(
                colors = colors,
                center = Offset(size.width * bg.centerX, size.height * bg.centerY)
            )
        }
        drawRect(brush)
        drawContent()
    }
}

/**
 * Apply pattern background (non-composable).
 */
private fun Modifier.applyStaticPattern(bg: Background.Pattern): Modifier {
    val primaryColor = parseColor(bg.primaryColor) ?: Color.Gray
    val secondaryColor = parseColor(bg.secondaryColor) ?: Color.LightGray
    
    return this.drawWithContent {
        // Draw base color
        drawRect(primaryColor)
        
        // Draw pattern
        when (bg.patternType) {
            PatternType.DOTS -> drawDotsPattern(secondaryColor, bg.size, bg.spacing)
            PatternType.STRIPES_HORIZONTAL -> drawHorizontalStripes(secondaryColor, bg.size, bg.spacing)
            PatternType.STRIPES_VERTICAL -> drawVerticalStripes(secondaryColor, bg.size, bg.spacing)
            PatternType.STRIPES_DIAGONAL -> drawDiagonalStripes(secondaryColor, bg.size, bg.spacing)
            PatternType.GRID -> drawGridPattern(secondaryColor, bg.size, bg.spacing)
            PatternType.CHECKERBOARD -> drawCheckerboard(secondaryColor, bg.size)
            PatternType.POLKA_DOTS -> drawPolkaDotsPattern(secondaryColor, bg.size, bg.spacing)
        }
        
        drawContent()
    }
}

/**
 * Apply image background.
 * Loads asynchronously via Coil and paints behind content using drawWithContent.
 */
@Composable
private fun Modifier.applyStaticImage(bg: Background.Image): Modifier {
    val context = LocalContext.current
    val imageLoaderFactory = LocalImageLoader.current
    val imageLoader = remember(context, imageLoaderFactory) {
        imageLoaderFactory?.invoke(context) ?: ImageLoaderProvider.getImageLoader(context)
    }

    val contentScale = when (bg.fit) {
        ImageFit.CROP -> ContentScale.Crop
        ImageFit.CONTAIN -> ContentScale.Fit
        ImageFit.FILL -> ContentScale.FillBounds
        ImageFit.TILE -> ContentScale.Crop
    }

    val painter = rememberAsyncImagePainter(
        model = ImageRequest.Builder(context)
            .data(bg.url)
            .crossfade(true)
            .build(),
        imageLoader = imageLoader,
        contentScale = contentScale
    )

    return this.drawWithContent {
        with(painter) {
            draw(size = this@drawWithContent.size, alpha = bg.opacity)
        }
        drawContent()
    }
}

// ============================================================================
// ANIMATED BACKGROUNDS (Composable)
// ============================================================================

/**
 * Apply shimmer effect background (animated, requires @Composable).
 */
@Composable
private fun Modifier.applyAnimatedShimmer(bg: Background.Shimmer): Modifier {
    val infiniteTransition = rememberInfiniteTransition(label = "shimmer")
    val offset by infiniteTransition.animateFloat(
        initialValue = -1f,
        targetValue = 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(bg.duration.toInt(), easing = LinearEasing),
            repeatMode = if (bg.loop) RepeatMode.Restart else RepeatMode.Reverse
        ),
        label = "shimmer_offset"
    )
    
    val baseColor = parseColor(bg.baseColor) ?: Color.LightGray
    val highlightColor = parseColor(bg.highlightColor) ?: Color.White
    
    return this.drawWithContent {
        val angleRadians = Math.toRadians((bg.angle - 90.0))
        // Offset pans the gradient across the component; scale by component dimension so
        // the sweep covers the full component regardless of its size.
        val panX = offset * size.width
        val panY = offset * size.height
        val shimmerStart = Offset(
            x = (0.5f - cos(angleRadians).toFloat() * 0.5f) * size.width + panX,
            y = (0.5f - sin(angleRadians).toFloat() * 0.5f) * size.height + panY
        )
        val shimmerEnd = Offset(
            x = (0.5f + cos(angleRadians).toFloat() * 0.5f) * size.width + panX,
            y = (0.5f + sin(angleRadians).toFloat() * 0.5f) * size.height + panY
        )
        val brush = Brush.linearGradient(
            colors = listOf(baseColor, highlightColor, baseColor),
            start = shimmerStart,
            end = shimmerEnd
        )
        drawRect(brush)
        drawContent()
    }
}

/**
 * Apply animated gradient background (animated, requires @Composable).
 */
@Composable
private fun Modifier.applyAnimatedGradient(bg: Background.AnimatedGradient): Modifier {
    val infiniteTransition = rememberInfiniteTransition(label = "animated_gradient")
    val animatedValue by infiniteTransition.animateFloat(
        initialValue = 0f,
        targetValue = 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(bg.duration.toInt(), easing = LinearEasing),
            repeatMode = if (bg.loop) RepeatMode.Restart else RepeatMode.Reverse
        ),
        label = "gradient_animation"
    )
    
    val colors = bg.colors.mapNotNull { parseColor(it) }
    if (colors.isEmpty()) return this
    
    return this.drawWithContent {
        // Rotate colors based on animation style
        val rotatedColors = when (bg.animationStyle) {
            AnimationStyle.SMOOTH -> colors  // Colors stay in place, just animate
            AnimationStyle.SHIFT -> {
                // Shift color positions
                val shift = (animatedValue * colors.size).toInt()
                colors.takeLast(shift) + colors.dropLast(shift)
            }
            AnimationStyle.PULSE -> colors  // Same as smooth for now
        }
        
        val angleRadians = Math.toRadians((bg.angle - 90.0))
        val gradStart = Offset(
            x = (0.5f - cos(angleRadians).toFloat() * 0.5f) * size.width,
            y = (0.5f - sin(angleRadians).toFloat() * 0.5f) * size.height
        )
        val gradEnd = Offset(
            x = (0.5f + cos(angleRadians).toFloat() * 0.5f) * size.width,
            y = (0.5f + sin(angleRadians).toFloat() * 0.5f) * size.height
        )
        val brush = when (bg.gradientType) {
            GradientType.LINEAR -> Brush.linearGradient(
                colors = rotatedColors,
                start = gradStart,
                end = gradEnd
            )
            GradientType.RADIAL -> Brush.radialGradient(
                colors = rotatedColors,
                center = Offset(size.width * 0.5f, size.height * 0.5f),
                radius = size.minDimension * (0.5f + animatedValue * 0.5f)  // Animate radius
            )
            GradientType.SWEEP -> Brush.sweepGradient(
                colors = rotatedColors,
                center = Offset(size.width * 0.5f, size.height * 0.5f)
            )
        }
        
        drawRect(brush)
        drawContent()
    }
}

/**
 * Apply pulse/breathing effect background (animated, requires @Composable).
 */
@Composable
private fun Modifier.applyAnimatedPulse(bg: Background.Pulse): Modifier {
    val infiniteTransition = rememberInfiniteTransition(label = "pulse")
    val alpha by infiniteTransition.animateFloat(
        initialValue = bg.minOpacity,
        targetValue = bg.maxOpacity,
        animationSpec = infiniteRepeatable(
            animation = tween(bg.duration.toInt(), easing = FastOutSlowInEasing),
            repeatMode = if (bg.loop) RepeatMode.Reverse else RepeatMode.Restart
        ),
        label = "pulse_alpha"
    )
    
    val color = parseColor(bg.color)?.copy(alpha = alpha) ?: Color.Transparent
    return this.background(color)
}

/**
 * Apply particles effect background (animated, requires @Composable).
 */
@Composable
private fun Modifier.applyAnimatedParticles(bg: Background.Particles): Modifier {
    val particles = remember {
        List(bg.particleCount) {
            Particle(
                x = Random.nextFloat(),
                y = Random.nextFloat(),
                vx = (Random.nextFloat() - 0.5f) * bg.speed,
                vy = when (bg.direction) {
                    ParticleDirection.UP -> -Random.nextFloat() * bg.speed
                    ParticleDirection.DOWN -> Random.nextFloat() * bg.speed
                    ParticleDirection.LEFT -> -Random.nextFloat() * bg.speed
                    ParticleDirection.RIGHT -> Random.nextFloat() * bg.speed
                    ParticleDirection.RANDOM -> (Random.nextFloat() - 0.5f) * bg.speed
                }
            )
        }
    }
    
    val infiniteTransition = rememberInfiniteTransition(label = "particles")
    val time by infiniteTransition.animateFloat(
        initialValue = 0f,
        targetValue = 1000f,
        animationSpec = infiniteRepeatable(
            animation = tween(1000000, easing = LinearEasing),
            repeatMode = RepeatMode.Restart
        ),
        label = "particle_time"
    )
    
    val particleColor = parseColor(bg.particleColor)?.copy(alpha = bg.opacity) ?: Color.White
    
    return this.drawWithContent {
        drawContent()
        
        particles.forEach { particle ->
            val x = ((particle.x + particle.vx * time * 0.001f) % 1f) * size.width
            val y = ((particle.y + particle.vy * time * 0.001f) % 1f) * size.height
            
            drawCircle(
                color = particleColor,
                radius = bg.particleSize,
                center = Offset(x, y)
            )
        }
    }
}

/**
 * Apply layered background (requires @Composable due to recursive calls).
 */
@Composable
private fun Modifier.applyLayeredBackground(bg: Background.Layered): Modifier {
    var modifier = this as Modifier
    bg.layers.forEach { layer ->
        modifier = modifier.applyBackground(layer)
    }
    return modifier
}

// ============================================================================
// HELPER FUNCTIONS (Non-Composable)
// ============================================================================


/**
 * Draw dots pattern.
 * Pure drawing function - no composition needed.
 */
private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawDotsPattern(
    color: Color,
    dotSize: Float,
    spacing: Float
) {
    val rows = (size.height / spacing).toInt() + 1
    val cols = (size.width / spacing).toInt() + 1
    
    for (row in 0..rows) {
        for (col in 0..cols) {
            drawCircle(
                color = color,
                radius = dotSize / 2,
                center = Offset(col * spacing, row * spacing)
            )
        }
    }
}

/**
 * Draw horizontal stripes.
 * Pure drawing function - no composition needed.
 */
private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawHorizontalStripes(
    color: Color,
    stripeHeight: Float,
    spacing: Float
) {
    val count = (size.height / (stripeHeight + spacing)).toInt() + 1
    for (i in 0..count) {
        val y = i * (stripeHeight + spacing)
        drawRect(
            color = color,
            topLeft = Offset(0f, y),
            size = Size(size.width, stripeHeight)
        )
    }
}

/**
 * Draw vertical stripes.
 * Pure drawing function - no composition needed.
 */
private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawVerticalStripes(
    color: Color,
    stripeWidth: Float,
    spacing: Float
) {
    val count = (size.width / (stripeWidth + spacing)).toInt() + 1
    for (i in 0..count) {
        val x = i * (stripeWidth + spacing)
        drawRect(
            color = color,
            topLeft = Offset(x, 0f),
            size = Size(stripeWidth, size.height)
        )
    }
}

/**
 * Draw diagonal stripes.
 * Pure drawing function - no composition needed.
 */
private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawDiagonalStripes(
    color: Color,
    stripeWidth: Float,
    spacing: Float
) {
    val count = ((size.width + size.height) / (stripeWidth + spacing)).toInt() + 1
    for (i in 0..count) {
        val offset = i * (stripeWidth + spacing)
        drawLine(
            color = color,
            start = Offset(offset, 0f),
            end = Offset(0f, offset),
            strokeWidth = stripeWidth
        )
    }
}

/**
 * Draw grid pattern.
 * Pure drawing function - no composition needed.
 */
private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawGridPattern(
    color: Color,
    lineWidth: Float,
    spacing: Float
) {
    drawHorizontalStripes(color, lineWidth, spacing)
    drawVerticalStripes(color, lineWidth, spacing)
}

/**
 * Draw checkerboard pattern.
 * Pure drawing function - no composition needed.
 */
private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawCheckerboard(
    color: Color,
    squareSize: Float
) {
    val rows = (size.height / squareSize).toInt() + 1
    val cols = (size.width / squareSize).toInt() + 1
    
    for (row in 0..rows) {
        for (col in 0..cols) {
            if ((row + col) % 2 == 0) {
                drawRect(
                    color = color,
                    topLeft = Offset(col * squareSize, row * squareSize),
                    size = Size(squareSize, squareSize)
                )
            }
        }
    }
}

/**
 * Draw polka dots pattern (larger dots with more spacing).
 * Pure drawing function - no composition needed.
 */
private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawPolkaDotsPattern(
    color: Color,
    dotSize: Float,
    spacing: Float
) {
    drawDotsPattern(color, dotSize, spacing)
}

/**
 * Particle data class.
 * Pure data structure - no composition needed.
 */
private data class Particle(
    val x: Float,
    val y: Float,
    val vx: Float,
    val vy: Float
)

File: ios/Sources/CleverTapNativeDisplay/Renderer/BackgroundRenderer.swift
// MARK: - Background Renderer
// Background rendering implementation for SwiftUI

import SwiftUI

/// View that renders a background based on the Background configuration.
struct BackgroundView: View {
    let background: Background

    init(background: Background) {
        self.background = background
    }

    var body: some View {
        switch background {
        case .solid(let bg):
            SolidBackgroundView(background: bg)
            
        case .linearGradient(let bg):
            LinearGradientBackgroundView(background: bg)
            
        case .radialGradient(let bg):
            RadialGradientBackgroundView(background: bg)
            
        case .sweepGradient(let bg):
            SweepGradientBackgroundView(background: bg)
            
        case .image(let bg):
            ImageBackgroundView(background: bg)
            
        case .shimmer(let bg):
            ShimmerBackgroundView(background: bg)
            
        case .animatedGradient(let bg):
            AnimatedGradientBackgroundView(background: bg)
            
        case .pulse(let bg):
            PulseBackgroundView(background: bg)
            
        case .pattern(let bg):
            PatternBackgroundView(background: bg)
            
        case .particles(let bg):
            ParticlesBackgroundView(background: bg)
            
        case .layered(let bg):
            LayeredBackgroundView(background: bg)
        }
    }
}

// MARK: - Static Backgrounds

/// Solid color background.
struct SolidBackgroundView: View {
    let background: SolidBackground
    
    var body: some View {
        ColorParser.parse(background.color) ?? Color.clear
    }
}

/// Linear gradient background.
struct LinearGradientBackgroundView: View {
    let background: LinearGradientBackground
    
    var body: some View {
        let colors = background.colors.compactMap { ColorParser.parse($0) }
        guard !colors.isEmpty else {
            return AnyView(Color.clear)
        }
        
        let stops: [Gradient.Stop]
        if let bgStops = background.stops, !bgStops.isEmpty {
            stops = zip(colors, bgStops).map { Gradient.Stop(color: $0.0, location: $0.1) }
        } else {
            stops = colors.enumerated().map { index, color in
                Gradient.Stop(color: color, location: CGFloat(index) / CGFloat(colors.count - 1))
            }
        }
        
        let (start, end) = calculateGradientPoints(angle: background.angle)
        
        return AnyView(
            LinearGradient(
                stops: stops,
                startPoint: start,
                endPoint: end
            )
        )
    }
    
    private func calculateGradientPoints(angle: CGFloat) -> (UnitPoint, UnitPoint) {
        let radians = (angle - 90) * .pi / 180
        let x = cos(radians)
        let y = sin(radians)
        
        let startX = 0.5 - x * 0.5
        let startY = 0.5 - y * 0.5
        let endX = 0.5 + x * 0.5
        let endY = 0.5 + y * 0.5
        
        return (
            UnitPoint(x: startX, y: startY),
            UnitPoint(x: endX, y: endY)
        )
    }
}

/// Radial gradient background.
struct RadialGradientBackgroundView: View {
    let background: RadialGradientBackground
    
    var body: some View {
        let colors = background.colors.compactMap { ColorParser.parse($0) }
        guard !colors.isEmpty else {
            return AnyView(Color.clear)
        }
        
        return AnyView(
            GeometryReader { geometry in
                let center = UnitPoint(x: background.centerX, y: background.centerY)
                let radius = min(geometry.size.width, geometry.size.height) * background.radius
                
                RadialGradient(
                    colors: colors,
                    center: center,
                    startRadius: 0,
                    endRadius: radius
                )
            }
        )
    }
}

/// Sweep/Angular gradient background.
struct SweepGradientBackgroundView: View {
    let background: SweepGradientBackground
    
    var body: some View {
        let colors = background.colors.compactMap { ColorParser.parse($0) }
        guard !colors.isEmpty else {
            return AnyView(Color.clear)
        }
        
        let center = UnitPoint(x: background.centerX, y: background.centerY)
        let startAngle = Angle(degrees: Double(background.startAngle))
        
        return AnyView(
            AngularGradient(
                colors: colors,
                center: center,
                startAngle: startAngle,
                endAngle: startAngle + .degrees(360)
            )
        )
    }
}

/// Image background.
struct ImageBackgroundView: View {
    let background: ImageBackground
    
    var body: some View {
        AsyncImage(url: URL(string: background.url)) { phase in
            switch phase {
            case .empty:
                Color.gray.opacity(0.2)
            case .success(let image):
                image
                    .resizable()
                    .aspectRatio(contentMode: contentMode)
                    .opacity(background.opacity)
                    .blur(radius: background.blur)
                    .overlay(tintOverlay)
            case .failure:
                Color.gray.opacity(0.3)
            @unknown default:
                Color.clear
            }
        }
    }
    
    private var contentMode: ContentMode {
        switch background.fit {
        case .crop: return .fill
        case .contain: return .fit
        case .fill: return .fill
        case .tile: return .fill // SwiftUI doesn't have native tiling
        }
    }
    
    @ViewBuilder
    private var tintOverlay: some View {
        if let tint = background.tint, background.tintOpacity > 0 {
            ColorParser.parse(tint)?
                .opacity(background.tintOpacity)
        }
    }
}

/// Pattern background.
struct PatternBackgroundView: View {
    let background: PatternBackground
    
    var body: some View {
        GeometryReader { geometry in
            let primaryColor = ColorParser.parse(background.primaryColor) ?? .gray
            let secondaryColor = ColorParser.parse(background.secondaryColor) ?? .white
            
            Canvas { context, size in
                // Draw base color
                context.fill(Path(CGRect(origin: .zero, size: size)), with: .color(primaryColor))
                
                // Draw pattern
                switch background.patternType {
                case .dots, .polkaDots:
                    drawDots(context: context, size: size, color: secondaryColor)
                case .stripesHorizontal:
                    drawHorizontalStripes(context: context, size: size, color: secondaryColor)
                case .stripesVertical:
                    drawVerticalStripes(context: context, size: size, color: secondaryColor)
                case .stripesDiagonal:
                    drawDiagonalStripes(context: context, size: size, color: secondaryColor)
                case .grid:
                    drawGrid(context: context, size: size, color: secondaryColor)
                case .checkerboard:
                    drawCheckerboard(context: context, size: size, color: secondaryColor)
                }
            }
            .opacity(background.opacity)
        }
    }
    
    private func drawDots(context: GraphicsContext, size: CGSize, color: Color) {
        let spacing = background.spacing
        let dotSize = background.size
        
        var y: CGFloat = 0
        while y < size.height + spacing {
            var x: CGFloat = 0
            while x < size.width + spacing {
                let path = Path(ellipseIn: CGRect(
                    x: x - dotSize / 2,
                    y: y - dotSize / 2,
                    width: dotSize,
                    height: dotSize
                ))
                context.fill(path, with: .color(color))
                x += spacing
            }
            y += spacing
        }
    }
    
    private func drawHorizontalStripes(context: GraphicsContext, size: CGSize, color: Color) {
        let stripeHeight = background.size
        let spacing = background.spacing
        
        var y: CGFloat = 0
        while y < size.height + spacing {
            let path = Path(CGRect(x: 0, y: y, width: size.width, height: stripeHeight))
            context.fill(path, with: .color(color))
            y += stripeHeight + spacing
        }
    }
    
    private func drawVerticalStripes(context: GraphicsContext, size: CGSize, color: Color) {
        let stripeWidth = background.size
        let spacing = background.spacing
        
        var x: CGFloat = 0
        while x < size.width + spacing {
            let path = Path(CGRect(x: x, y: 0, width: stripeWidth, height: size.height))
            context.fill(path, with: .color(color))
            x += stripeWidth + spacing
        }
    }
    
    private func drawDiagonalStripes(context: GraphicsContext, size: CGSize, color: Color) {
        let stripeWidth = background.size
        let spacing = background.spacing
        let diagonal = sqrt(size.width * size.width + size.height * size.height)
        
        var offset: CGFloat = -diagonal
        while offset < diagonal {
            var path = Path()
            path.move(to: CGPoint(x: offset, y: 0))
            path.addLine(to: CGPoint(x: offset + diagonal, y: diagonal))
            context.stroke(path, with: .color(color), lineWidth: stripeWidth)
            offset += stripeWidth + spacing
        }
    }
    
    private func drawGrid(context: GraphicsContext, size: CGSize, color: Color) {
        drawHorizontalStripes(context: context, size: size, color: color)
        drawVerticalStripes(context: context, size: size, color: color)
    }
    
    private func drawCheckerboard(context: GraphicsContext, size: CGSize, color: Color) {
        let squareSize = background.size
        
        var row = 0
        var y: CGFloat = 0
        while y < size.height + squareSize {
            var col = 0
            var x: CGFloat = 0
            while x < size.width + squareSize {
                if (row + col) % 2 == 0 {
                    let path = Path(CGRect(x: x, y: y, width: squareSize, height: squareSize))
                    context.fill(path, with: .color(color))
                }
                x += squareSize
                col += 1
            }
            y += squareSize
            row += 1
        }
    }
}

// MARK: - Animated Backgrounds

/// Shimmer effect background.
struct ShimmerBackgroundView: View {
    let background: ShimmerBackground
    @State private var offset: CGFloat = -1
    
    var body: some View {
        let baseColor = ColorParser.parse(background.baseColor) ?? .gray
        let highlightColor = ColorParser.parse(background.highlightColor) ?? .white
        
        GeometryReader { geometry in
            let (start, end) = calculateGradientPoints(angle: background.angle, offset: offset)
            
            LinearGradient(
                colors: [baseColor, highlightColor, baseColor],
                startPoint: start,
                endPoint: end
            )
        }
        .onAppear {
            if background.loop {
                withAnimation(.linear(duration: Double(background.duration) / 1000).repeatForever(autoreverses: false)) {
                    offset = 1
                }
            } else {
                withAnimation(.linear(duration: Double(background.duration) / 1000)) {
                    offset = 1
                }
            }
        }
    }
    
    private func calculateGradientPoints(angle: CGFloat, offset: CGFloat) -> (UnitPoint, UnitPoint) {
        let radians = (angle - 90) * .pi / 180
        let x = cos(radians)
        let y = sin(radians)
        
        let startX = 0.5 - x * 0.5 + offset * x
        let startY = 0.5 - y * 0.5 + offset * y
        let endX = 0.5 + x * 0.5 + offset * x
        let endY = 0.5 + y * 0.5 + offset * y
        
        return (
            UnitPoint(x: startX, y: startY),
            UnitPoint(x: endX, y: endY)
        )
    }
}

/// Animated gradient background.
struct AnimatedGradientBackgroundView: View {
    let background: AnimatedGradientBackground
    @State private var animationProgress: CGFloat = 0
    
    var body: some View {
        let colors = background.colors.compactMap { ColorParser.parse($0) }
        guard !colors.isEmpty else {
            return AnyView(Color.clear)
        }
        
        return AnyView(
            TimelineView(.animation) { timeline in
                let phase = timeline.date.timeIntervalSince1970.truncatingRemainder(dividingBy: Double(background.duration) / 1000)
                let progress = phase / (Double(background.duration) / 1000)
                
                gradientView(colors: colors, progress: CGFloat(progress))
            }
        )
    }
    
    @ViewBuilder
    private func gradientView(colors: [Color], progress: CGFloat) -> some View {
        switch background.gradientType {
        case .linear:
            let (start, end) = animatedLinearPoints(progress: progress)
            LinearGradient(colors: colors, startPoint: start, endPoint: end)
            
        case .radial:
            let radius = 0.5 + progress * 0.5
            GeometryReader { geometry in
                RadialGradient(colors: colors, center: .center, startRadius: 0, endRadius: geometry.size.width * radius)
            }
            
        case .sweep:
            let startAngle = Angle(degrees: Double(progress * 360))
            AngularGradient(colors: colors, center: .center, startAngle: startAngle, endAngle: startAngle + .degrees(360))
        }
    }
    
    private func animatedLinearPoints(progress: CGFloat) -> (UnitPoint, UnitPoint) {
        let angle = background.angle + progress * 360
        let radians = angle * .pi / 180
        let x = cos(radians)
        let y = sin(radians)
        
        return (
            UnitPoint(x: 0.5 - x * 0.5, y: 0.5 - y * 0.5),
            UnitPoint(x: 0.5 + x * 0.5, y: 0.5 + y * 0.5)
        )
    }
}

/// Pulse/breathing effect background.
struct PulseBackgroundView: View {
    let background: PulseBackground
    @State private var opacity: Double = 0.3
    
    var body: some View {
        let color = ColorParser.parse(background.color) ?? .blue
        
        color
            .opacity(opacity)
            .onAppear {
                let duration = Double(background.duration) / 1000
                withAnimation(
                    .easeInOut(duration: duration)
                    .repeatForever(autoreverses: true)
                ) {
                    opacity = background.maxOpacity
                }
            }
    }
}

/// Particles effect background.
struct ParticlesBackgroundView: View {
    let background: ParticlesBackground
    @State private var particles: [Particle] = []
    @State private var time: Date = Date()
    
    private struct Particle: Identifiable {
        let id = UUID()
        var x: CGFloat
        var y: CGFloat
        var vx: CGFloat
        var vy: CGFloat
    }
    
    var body: some View {
        let particleColor = ColorParser.parse(background.particleColor)?.opacity(background.opacity) ?? .white.opacity(0.7)
        
        TimelineView(.animation) { context in
            Canvas { ctx, size in
                let elapsed = context.date.timeIntervalSince(time)
                
                for particle in particles {
                    let x = ((particle.x + particle.vx * CGFloat(elapsed) * 0.01).truncatingRemainder(dividingBy: 1) + 1).truncatingRemainder(dividingBy: 1) * size.width
                    let y = ((particle.y + particle.vy * CGFloat(elapsed) * 0.01).truncatingRemainder(dividingBy: 1) + 1).truncatingRemainder(dividingBy: 1) * size.height
                    
                    let path = Path(ellipseIn: CGRect(
                        x: x - background.particleSize / 2,
                        y: y - background.particleSize / 2,
                        width: background.particleSize,
                        height: background.particleSize
                    ))
                    ctx.fill(path, with: .color(particleColor))
                }
            }
        }
        .onAppear {
            time = Date()
            particles = (0..<background.particleCount).map { _ in
                let vy: CGFloat
                switch background.direction {
                case .up: vy = -CGFloat.random(in: 0..<background.speed)
                case .down: vy = CGFloat.random(in: 0..<background.speed)
                case .left: vy = 0
                case .right: vy = 0
                case .random: vy = CGFloat.random(in: -background.speed..<background.speed)
                }
                
                let vx: CGFloat
                switch background.direction {
                case .up, .down: vx = CGFloat.random(in: -0.5..<0.5) * background.speed
                case .left: vx = -CGFloat.random(in: 0..<background.speed)
                case .right: vx = CGFloat.random(in: 0..<background.speed)
                case .random: vx = CGFloat.random(in: -background.speed..<background.speed)
                }
                
                return Particle(
                    x: CGFloat.random(in: 0..<1),
                    y: CGFloat.random(in: 0..<1),
                    vx: vx,
                    vy: vy
                )
            }
        }
    }
}

/// Layered background.
struct LayeredBackgroundView: View {
    let background: LayeredBackground
    
    var body: some View {
        ZStack {
            ForEach(background.layers.indices, id: \.self) { index in
                BackgroundView(background: background.layers[index])
            }
        }
    }
}
